scadnano/scadnano.py

```diff
@@ -4024,22 +4024,35 @@ def _cadnano_v2_import_find_5_end(vstrands: VStrands, strand_type: str, helix_nu
         """ Routine which finds the 5' end of a strand in a cadnano v2 import. It returns the
         helix and the base of the 5' end.
         """
-        id_from_before = helix_num
+        id_from_before = helix_num # 'id' stands for helix id
         base_from_before = base_id
+
+        circular_seen = {}
+        is_circular = False
+
         while not (id_from == -1 and base_from == -1):
+            if (id_from,base_from) in circular_seen:
+                is_circular = True
+                break
+            circular_seen[(id_from,base_from)] = True
             id_from_before = id_from
             base_from_before = base_from
             id_from, base_from, _, _ = vstrands[id_from][strand_type][base_from]
-        return id_from_before, base_from_before
+        return id_from_before, base_from_before, is_circular

```

Comment:
The return type of this method should be changed to `Tuple[int, int, bool]`. (Run mypy from the command line to see a list of mypy errors, or open in PyCharm.) There are some mypy errors in other parts of the code (not dealing with cadnano import/export) that are fixed in a newer branch I have, so don't worry about those errors, but here's the mypy errors in code related to cadnano import/export:

```
scadnano.py:4041: error: Incompatible return value type (got "Tuple[int, int, bool]", expected "Tuple[int, int]")
scadnano.py:4149: error: "Sequence[Domain[Any]]" has no attribute "__delitem__"; maybe "__getitem__"?
scadnano.py:4166: error: Need more than 2 values to unpack (3 expected)
scadnano.py:4174: error: Cannot determine type of 'strand_5_end_base'
scadnano.py:4175: error: Cannot determine type of 'strand_5_end_helix'
scadnano.py:4177: error: Cannot determine type of 'strand_5_end_base'
scadnano.py:4178: error: Cannot determine type of 'strand_5_end_helix'
scadnano.py:4180: error: Cannot determine type of 'is_circular'
scadnano.py:4185: error: Cannot determine type of 'is_circular'
scadnano.py:4454: error: Item "Loopout" of "Union[Domain[Any], Loopout]" has no attribute "helix"
scadnano.py:4455: error: Item "Loopout" of "Union[Domain[Any], Loopout]" has no attribute "start"
scadnano.py:4455: error: Item "Loopout" of "Union[Domain[Any], Loopout]" has no attribute "end"
scadnano.py:4455: error: Item "Loopout" of "Union[Domain[Any], Loopout]" has no attribute "forward"
scadnano.py:4458: error: Item "Loopout" of "Union[Domain[Any], Loopout]" has no attribute "helix"
scadnano.py:4459: error: Item "Loopout" of "Union[Domain[Any], Loopout]" has no attribute "start"
scadnano.py:4459: error: Item "Loopout" of "Union[Domain[Any], Loopout]" has no attribute "end"
scadnano.py:4459: error: Item "Loopout" of "Union[Domain[Any], Loopout]" has no attribute "forward"
```